### PR TITLE
ohbase table group route

### PR DESCRIPTION
**Pull request by foronedream**
ohbase table group search 
## Summary

ohbase table group search 

## Solution Description

ohbase table group search

### CODE REVIEW

**Comment by IHEII** [comment]
I think we could do the transform of table name [here](https://github.com/oceanbase/obkv-table-client-java/blob/fdfe9fbc51fb2e51abdb7a88950b42b56478e7b3/src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java#L2572). After we transform a table group name into a table name. We just go through the normal procedure.

File: src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java [hunk 1]
```diff
@@ -156,6 +156,11 @@ public class ObTableClient extends AbstractObTableClient implements Lifecycle {
     private int                                               odpPort                                 = 2883;
 
     private ObTable                                           odpTable                                = null;
+    // tableGroup <-> Table
+    private ConcurrentHashMap<String, Lock>                   TableGroupCacheLocks                    = new ConcurrentHashMap<String, Lock>();
+    private ConcurrentHashMap<String, String>                 TableGroupCache                         = new ConcurrentHashMap<String, String>();  // tableGroup -> Table
+    private ConcurrentHashMap<String, String>                 TableGroupInverted                      = new ConcurrentHashMap<String, String>();  // Table -> tableGroup
+    private boolean                                           isTableGroup                            = false;
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java [hunk 2]
```diff
@@ -125,24 +136,40 @@ protected ObTableQueryResult execute(ObPair<Long, ObTableParam> partIdWithIndex,
...
-                               && ((ObTableException) e).isNeedRefreshTableEntry()) {
-                        needRefreshTableEntry = true;
-                        logger
-                            .warn(
-                                "tablename:{} partition id:{} stream query refresh table while meet Exception needing refresh, errorCode: {}",
-                                indexTableName, partIdWithIndex.getLeft(),
-                                ((ObTableException) e).getErrorCode(), e);
-                        if (client.isRetryOnChangeMasterTimes()
-                            && (tryTimes - 1) < client.getRuntimeRetryTimes()) {
+                    } else if (e instanceof ObTableException) {
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java:142]: edit

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java [hunk 3]
```diff
@@ -125,24 +136,40 @@ protected ObTableQueryResult execute(ObPair<Long, ObTableParam> partIdWithIndex,
...
+                        if (((ObTableException) e).getErrorCode() == ResultCodes.OB_TABLE_NOT_EXIST.errorCode
+                                && ((ObTableQueryRequest) request).getTableQuery().isHbaseQuery() && client.isTableGroup()) {
+                            if (client.getTableGroupInverted().get(indexTableName) == null) {
+                                throw new ObTableNotExistException("tableGroup not exists, TableName = " + indexTableName);
+                            } else {
+                                indexTableName = client.tryGetTableNameFromTableGroupCache(client.getTableGroupInverted().get(indexTableName), true);
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/table/ObTableClientQueryImpl.java [hunk 4]
```diff
@@ -137,6 +137,10 @@ public ObTableClientQueryStreamResult executeInternal() throws Exception {
             tableQuery.addKeyRange(ObNewRange.getWholeRange());
         }
         if (obTableClient.isOdpMode()) {
+            // TableGroup -> TableName
+            if (tableQuery.isHbaseQuery() && tableQuery.isTableGroupName(tableName)) {
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/table/ObTableClientQueryImpl.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java [hunk 5]
```diff
@@ -2570,13 +2690,25 @@ public ObPayload execute(final ObTableAbstractOperationRequest request) throws E
                 .executeInternal();
             return batchOpsResult.getResults().get(0);
         } else if (request instanceof ObTableQueryRequest) {
-            ObTableClientQueryImpl tableQuery = new ObTableClientQueryImpl(request.getTableName(),
+            // TableGroup -> TableName
+            String tableName = request.getTableName();
+            if (((ObTableQueryRequest) request).getTableQuery().isHbaseQuery() 
+                    && isTableGroupName(tableName)) {
+                tableName = tryGetTableNameFromTableGroupCache(tableName, false);
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java:2697]: When will you replace the realTableName with tableGroupName

File: src/test/java/com/alipay/oceanbase/rpc/hbase/ObHTableTest.java [hunk 6]
```diff
@@ -30,6 +32,7 @@
 
 public class ObHTableTest {
     private ObTable client;
+    private ObTableClient obTableClient;
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/test/java/com/alipay/oceanbase/rpc/hbase/ObHTableTest.java:35]: hbase mode set in request

File: src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java (hunk 5 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java:2697]: image

File: src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java [hunk 7]
```diff
@@ -1285,6 +1289,49 @@ private TableEntry refreshTableEntry(TableEntry tableEntry, String tableName, bo
...
     }
 
+    /**
+     * 根据 tableGroup 获取其中一个tableName
+     * @param realTableName
+     * @param tableName
+     * @return
+* @throws Exception
+     */
+    private String refreshTableName(String realTableName, String tableName) throws Exception {
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java [hunk 8]
```diff
@@ -1285,6 +1289,49 @@ private TableEntry refreshTableEntry(TableEntry tableEntry, String tableName, bo
         return tableEntry;
     }
 
+    /**
+     * 根据 tableGroup 获取其中一个tableName
+     * @param realTableName
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java [hunk 9]
```diff
@@ -1679,6 +1726,79 @@ public List<ObPair<Long, ObTableParam>> getTables(String tableName, Object[] sta
         return obTableParams;
     }
 
+    /**
+     * get table name with table group
+     * @param tableName
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java [hunk 10]
```diff
@@ -437,6 +441,155 @@ TableEntry execute(Connection connection)
...
+            }
+        }
+
+        if (realTableName != null && !realTableName.isEmpty()) {
+            return realTableName;
+        } else {
+            throw new ObTableNotExistException("table name is invalid, addr = " + obServerAddr
+                    + " key =" + key + " tableName =" + realTableName);
+        }
+
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java [hunk 11]
```diff
@@ -437,6 +441,155 @@ TableEntry execute(Connection connection)
...
+        } catch (Throwable t) {
+            RUNTIME.error("callTableEntryNameWithPriority meet exception", t);
+            throw t;
+        }
+    }
+
+    /*
+     * load Table Name With table Group
+     */
+    public static String loadTableNameWithGroup(final ServerRoster serverRoster,
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java [hunk 12]
```diff
@@ -125,24 +138,38 @@ protected ObTableQueryResult execute(ObPair<Long, ObTableParam> partIdWithIndex,
...
-                        needRefreshTableEntry = true;
-                        logger
-                            .warn(
-                                "tablename:{} partition id:{} stream query refresh table while meet Exception needing refresh, errorCode: {}",
-                                indexTableName, partIdWithIndex.getLeft(),
-                                ((ObTableException) e).getErrorCode(), e);
-                        if (client.isRetryOnChangeMasterTimes()
-                            && (tryTimes - 1) < client.getRuntimeRetryTimes()) {
+                    } else if (e instanceof ObTableException) {
+                        if ((((ObTableException) e).getErrorCode() == ResultCodes.OB_TABLE_NOT_EXIST.errorCode
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java [hunk 13]
```diff
@@ -125,24 +138,38 @@ protected ObTableQueryResult execute(ObPair<Long, ObTableParam> partIdWithIndex,
...
+                                    || ((ObTableException) e).getErrorCode() == ResultCodes.OB_NOT_SUPPORTED.errorCode)
+                                && ((ObTableQueryRequest) request).getTableQuery().isHbaseQuery()) {
+                            if (client.getTableGroupInverted().get(indexTableName) == null) {
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java]: edit, not support report error

File: src/test/java/com/alipay/oceanbase/rpc/hbase/ObHTableTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (IHEII) [reply src/test/java/com/alipay/oceanbase/rpc/hbase/ObHTableTest.java:38]: client.setRunningMode()

File: src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java [hunk 14]
```diff
@@ -437,6 +441,104 @@ TableEntry execute(Connection connection)
...
+        } else {
+            throw new ObTableNotExistException("table name is invalid, addr = " + addr + " key =" + key + " tableName =" + realTableName);
+        }
+        
+    }
+
+    /*
+     * get TableName From Remote with Group
+     */
+    private static String getTableNameFromRemote(Connection connection, TableEntryKey key)
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java:494]: getTableNameByGroupNameFromRemote

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java [hunk 15]
```diff
@@ -76,10 +80,20 @@ protected ObTableQueryResult execute(ObPair<Long, ObTableParam> partIdWithIndex,
...
+                        if (isNeedRefreshTableGroupName) {
+                            // if input is table group, retry should replace table name
+                            indexTableName = client.tryGetTableNameFromTableGroupCache(client.getTableGroupInverted().get(indexTableName), true);
+                            tableName = indexTableName;
+                        }
+                        tableParam = client.getTable(indexTableName, partIdWithIndex.getLeft(),
+                                        needRefreshTableEntry, client.isTableEntryRefreshIntervalWait(),
+                                        route).getRight();
+                        subObTable = tableParam.getObTable();
+                        if (isNeedRefreshTableGroupName) {
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java:92]: Reset isNeedRefreshTableGroupName
Also, whether it needs to refresh the tableGroup name, we could set new tableId and tabletId. This means we could remove `if`

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java [hunk 16]
```diff
@@ -125,21 +139,33 @@ protected ObTableQueryResult execute(ObPair<Long, ObTableParam> partIdWithIndex,
...
-                               && ((ObTableException) e).isNeedRefreshTableEntry()) {
-                        needRefreshTableEntry = true;
-                        logger
-                            .warn(
-                                "tablename:{} partition id:{} stream query refresh table while meet Exception needing refresh, errorCode: {}",
-                                indexTableName, partIdWithIndex.getLeft(),
-                                ((ObTableException) e).getErrorCode(), e);
-                        if (client.isRetryOnChangeMasterTimes()
-                            && (tryTimes - 1) < client.getRuntimeRetryTimes()) {
+                    } else if (e instanceof ObTableException) {
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java:142]: OB_TABLE_NOT_EXIST is an ObTableEntryRefreshException. You could use the previous code to enter this block. And deal with table group inside the previous block

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java [hunk 17]
```diff
@@ -76,10 +80,20 @@ protected ObTableQueryResult execute(ObPair<Long, ObTableParam> partIdWithIndex,
                         if (failedServerList != null) {
                             route.setBlackList(failedServerList);
                         }
-                        subObTable = client
-                            .getTable(indexTableName, partIdWithIndex.getLeft(),
-                                needRefreshTableEntry, client.isTableEntryRefreshIntervalWait(),
-                                route).getRight().getObTable();
+                        if (isNeedRefreshTableGroupName) {
+                            // if input is table group, retry should replace table name
+                            indexTableName = client.tryGetTableNameFromTableGroupCache(client.getTableGroupInverted().get(indexTableName), true);
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java:85]: We could ignore indexTableName here. It seems that we use real TableName when retry.
Also, we could use 
```
tableName = client.tryGetTableNameFromTableGroupCache(client.getTableGroupInverted().get(indexTableName), isNeedRefreshTableGroupName;
```

File: src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java [hunk 18]
```diff
@@ -437,6 +441,104 @@ TableEntry execute(Connection connection)
             });
     }
 
+    /*
+     * load Table Name With table Group
+     */
+    public static String loadTableNameWithGroup(final ServerRoster serverRoster,
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java [hunk 19]
```diff
@@ -1285,6 +1289,49 @@ private TableEntry refreshTableEntry(TableEntry tableEntry, String tableName, bo
         return tableEntry;
     }
 
+    /**
+     * 根据 tableGroup 获取其中一个tableName
+     * @param physicalTableName // Complete table from table group
```
[the remark this answers is not in the document]
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java (hunk 14 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/location/LocationUtil.java]: edit

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java (hunk 15 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java]: edit , delete retry get table name

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java (hunk 16 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java:130]: edit , clear table group cache

File: src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java (hunk 17 above)
Comment (foronedream) [reply src/main/java/com/alipay/oceanbase/rpc/stream/ObTableClientQueryStreamResult.java]: edit, delete

File: src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java [hunk 20]
```diff
@@ -3086,4 +3219,31 @@ public String toString() {
...
+    public ConcurrentHashMap<String, String> getTableGroupCache() {
+        return TableGroupCache;
+    }
+
+    /**
+     * get table route fail than clear table group message
+     * @param phyTableName
+     * @param tableGroupName
+     */
+    public void clearTableGroupCache(String phyTableName, String tableGroupName) {
```
Comment (IHEII) [inline src/main/java/com/alipay/oceanbase/rpc/ObTableClient.java:3235]: tableGroupName is enough
BTW, 'clean'/'delete'/'erase' are better than 'clear'
`eraseTableGroupFromCache(String tableGroupName);`

File: src/test/java/com/alipay/oceanbase/rpc/hbase/ObHTableTest.java [hunk 21]
```diff
@@ -90,4 +98,110 @@ PRIMARY KEY (`K`, `Q`, `T`)
...
+
+        ObHTableOperationRequest hTableOperationRequestScan2 = new ObHTableOperationRequest();
+        hTableOperationRequestScan2.setOperationType(ObTableOperationType.SCAN);
+        hTableOperationRequestScan2.setTableName("test");
+        hTableOperationRequestScan2.setRowKey("putKey".getBytes());
+
+        ObTableQueryRequest requestScan2 = (ObTableQueryRequest)hTableOperationRequestScan2.obTableGroupOperationRequest();
+        ObTableClientQueryStreamResult clientQueryStreamResultScan2 = (ObTableClientQueryStreamResult) obTableClient
+                .execute(requestScan2);
+        //getKeyValueFromResult(clientQueryStreamResultScan2, true, family);
```
Comment (IHEII) [inline src/test/java/com/alipay/oceanbase/rpc/hbase/ObHTableTest.java:203]: Remove useless comment

**Review by IHEII — COMMENTED** [review]
Try to pull the latest code

**Pull request closed without merging**